### PR TITLE
KMP

### PR DESCRIPTION
**Pull request by rspencer01**

### CODE REVIEW

File: KMP/Test.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+public class Test
```
Comment (h4tguy) [inline KMP/Test.java]: Main

File: KMP/KMP.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+public class KMP
+{
+  private static int[] table(String n)
+  {
+    int[] T = new int[n.length()];
```
Comment (h4tguy) [inline KMP/KMP.java]: What is T,p,c etc etc

File: KMP/KMP.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+public class KMP
+{
+  private static int[] table(String n)
```
Comment (h4tguy) [inline KMP/KMP.java]: I think maybe it is better to do this on arrays, but it is easy enough to convert.

File: KMP/KMP.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    {
+      if (n.charAt(p-1) == n.charAt(c))
+      {
+        c++;
+        T[p] = c;
+        p++;
+      }
+      else
+      {
+        if (c>0)
```
Comment (h4tguy) [inline KMP/KMP.java]: if the else has braces, then so should the if.

File: KMP/KMP.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  //     be found ending at i in n
+  // p : The position we are computing in T
+  // c : The index of the next candidate to be
+  //     added to the substring we are considering
+  private static int[] table(String n) {
+    int[] T = new int[n.length()];
+    int p = 2;
+    int c = 0;
+    T[0] = -1;
+    if (n.length()==1)
```
Comment (h4tguy) [inline KMP/KMP.java]: <=

File: KMP/KMP.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      }
+    }
+    return T;
+  }
+  // n : The needle
+  // h : The haystack
+  // T : The table of lookbacks
+  // m : The position in the haystack we are checking
+  // i : The length into the needle we are checking
+  public static int KMP(String n, String h) {
```
Comment (h4tguy) [inline KMP/KMP.java]: n.length == 0 ?

Comment (rspencer01) [inline KMP/KMP.java]: Mmm.  Yes.  Return 0 or -1?

File: KMP/KMP.java (hunk 5 above)
Comment (rspencer01) [inline KMP/KMP.java]: Not sure what you are saying here.  Do you mean the `n.length()==0` case?  If so, it is sorted now.

**Comment by h4tguy** [comment]
lgtm. Feel free to merge.

**Comment by rspencer01** [comment]
@h4tguy Go for cpp?

**Comment by h4tguy** [comment]
Oh, didn't see that.

**Comment by h4tguy** [comment]
We might care about finding all matches, in which the table needs to generate one more entry (if we match the whole string, where do we go?), but otherwise lgtm.

**Pull request merged**